cipher_reactivity/cipher_reactivity/trigger.py

```diff
@@ -0,0 +1,136 @@
+    ReactionNode,
+    Retrosynthesis,
+    SyntheticTree,
+)
+from cipher_reactivity.sascorer import SAScorer
+
+me.connect(host=os.environ["MONGO_URI"])
+MONGO_CLIENT = pmg.MongoClient(os.environ["MONGO_URI"])
+
+ASKCOST_HOST = os.environ["ASKCOS_HOST"]
+try:
+    TREE_PARAMS = tomli.loads(resources.read_text("cipher_reactivity.data", "tree_params.toml"))
+except ValueError:
+    TREE_PARAMS = None
+ASKCOS_CLIENT = AskcosClient(ASKCOST_HOST, TREE_PARAMS)
+
+SA_SCORER = SAScorer(json.loads(resources.read_text("cipher_reactivity.data", "fpscores.json")))
+
+
+def build_rxn_node(tree: Dict) -> ReactionNode:
+    rn = ReactionNode()
+
+    rn.smiles = tree["smiles"]
+    rn.tforms = tree["tforms"]
+    rn.tsources = tree["tsources"]
+    rn.template_score = tree["template_score"]
+    rn.plausiblity = tree["plausibility"]
+    rn.rank = tree["rank"]
+    rn.num_examples = tree["num_examples"]
+    rn.necessary_reagent = tree["necessary_reagent"]
+    rn.precursor_smiles = tree["precursor_smiles"]
+    rn.rms_molwt = tree["rms_molwt"]
+    rn.num_rings = tree["num_rings"]
+    rn.scscore = tree["scscore"]
+    rn.rxn_id = tree["id"]
+    rn.forward_score = tree["forward_score"]
+    rn.class_num = tree["class_num"]
+    rn.class_name = tree["class_name"]
+    rn.children = [build_chemical_node(child) for child in tree["children"]]
+
+    return rn
+
+
+def build_chemical_node(tree: Dict) -> ChemicalNode:
+    cn = ChemicalNode()
+
+    cn.smiles = tree["smiles"]
+    cn.ppg = tree["ppg"]
+    cn.as_product = tree["as_product"]
+    cn.as_reactant = tree["as_reactant"]
+    cn.terminal = tree["terminal"]
+    cn.chemical_id = tree["id"]
+    if not cn.terminal:
+        cn.rxn = build_rxn_node(tree["children"][0])
+
+    return cn
+
+
+def build_synthetic_tree(tree: Dict) -> SyntheticTree:
+    st = SyntheticTree()
+
+    st.depth = tree["attributes"]["depth"]
+    st.precursor_cost = tree["attributes"]["precursor_cost"]
+    st.score = tree["attributes"]["score"]
+    st.cluster_id = tree["attributes"]["cluster_id"]
+    st.root = build_chemical_node(tree)
+
+    with tempfile.NamedTemporaryFile() as fid, warnings.catch_warnings():
+        warnings.simplefilter("ignore", category=graphviz.exceptions.UnknownSuffixWarning)
+        g = viz.tree_to_graph(viz.clean_tree(tree))
+        g.render(outfile=fid.name, format="png", cleanup=True)
+        fid.seek(0)
+        st.image = fid
+
+    return st
+
+
+def update_retrosynthesis(inchikey, smi):
+    retro = Retrosynthesis()
+
+    retro.inchikey = inchikey
+    retro.smi = smi
+    retro.trees = [build_synthetic_tree(tree) for tree in ASKCOS_CLIENT.get_trees(smi)]
+
+    retro.save()
+
+    return retro
+
+
+def update_difficulty(inchikey, smi):
+    difficulty = Difficulty()
+
+    difficulty.inchikey = inchikey
+    difficulty.smiles = smi
+    difficulty.sc_score = ASKCOS_CLIENT.sc_score(smi)
+    difficulty.sa_score = SA_SCORER(smi)
+
+    difficulty.save()
+
+    return difficulty
```

Comment:
I think we need to have a way to build a reaction tree outside of the trigger. For example if we do a mass update on the database and the retrosynthesis trigger is not active the retrosynthetic trees would need to be retroactively updated for all of those compound added. Zack and I put ours in the util folders along with all methods for updating the database, but I think this is fine as well as long as the utility exists somewhere